signer/src/network/libp2p/bootstrap.rs

```diff
@@ -0,0 +1,314 @@
+    {
+        // If we have any pending events, we return them immediately.
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+
+        // If we have any connections, we consider ourselves bootstrapped.
+        // Determine if there was a state change or not and return the correct
+        // poll result.
+        if self.connection_count >= 1 {
+            return match self.is_bootstrapped {
+                true => Poll::Pending,
+                false => {
+                    // We've just bootstrapped, so we generate an event to notify the
+                    // swarm.
+                    tracing::debug!(connection_count = %self.connection_count, "bootstrapping complete");
+                    self.is_bootstrapped = true;
+                    Poll::Ready(ToSwarm::GenerateEvent(BootstrapEvent::Complete))
+                }
+            };
+        }
+
+        // If we're here then we're not connected to any peers. If our current
+        // state is bootstrapped, we need to re-bootstrap.
+        if self.is_bootstrapped {
+            tracing::debug!(
+                "state is bootstrapped but not connected to any peers; re-bootstrapping is needed"
+            );
+            self.is_bootstrapped = false;
+            self.pending_events
+                .push_back(ToSwarm::GenerateEvent(BootstrapEvent::Needed));
+        }
+
+        // If we've attempted to bootstrap recently, we wait until the interval
+        // has passed.
+        if let Some(last_bootstrap) = self.last_attempted_at {
+            if last_bootstrap.elapsed() < self.config.bootstrap_interval {
+                return Poll::Pending;
+            }
+        }
+
+        // Queue the bootstrap started event.
+        tracing::debug!(addresses = ?self.config.seed_addresses, "initiating network bootstrapping from seed addresses");
+        self.pending_events
+            .push_back(ToSwarm::GenerateEvent(BootstrapEvent::Started {
+                addresses: self.config.seed_addresses.clone(),
+            }));
+
+        // Iterate over the seed addresses and queue dial events for each. Note
+        // that we queue a dial event for each seed address, regardless of our
+        // current connection count (which will then be propagated in subsequent
+        // polls). This is to ensure the best chance of connecting to the
+        // network.
+        self.config.seed_addresses.iter().for_each(|addr| {
+            // Construct the dialing options and `Dial` event which will be
+            // sent to the swarm.
+            let dial_opts = DialOpts::unknown_peer_id().address(addr.clone()).build();
+            let event = ToSwarm::Dial { opts: dial_opts };
+
+            // Queue the dial event.
+            self.pending_events.push_back(event);
+        });
+
+        // Update the last bootstrap attempt time.
+        self.last_attempted_at = Some(Instant::now());
+
+        Poll::Pending
```

Comment:
qq: why not popping and returning the head of `self.pending_events` here directly?

Reply:
Good point, [f87c47e](https://github.com/stacks-network/sbtc/pull/1257/commits/f87c47ee29ba6d110a49fd7a28977f51b24035b1)